casper/src/main/scala/coop/rchain/casper/blocks/proposer/BlockCreator.scala

```diff
@@ -0,0 +1,181 @@
+import coop.rchain.casper.util.rholang._
+import coop.rchain.casper.util.rholang.costacc.{CloseBlockDeploy, SlashDeploy}
+import coop.rchain.casper.{CasperSnapshot, PrettyPrinter, ValidatorIdentity}
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.signatures.Signed
+import coop.rchain.metrics.{Metrics, Span}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.rholang.interpreter.SystemProcesses.BlockData
+import coop.rchain.shared.{Log, Time}
+
+object BlockCreator {
+  private[this] val ProcessDeploysAndCreateBlockMetricsSource =
+    Metrics.Source(Metrics.BaseSource, "create-block")
+
+  /*
+   * Overview of createBlock
+   *
+   *  1. Rank each of the block cs's latest messages (blocks) via the LMD GHOST estimator.
+   *  2. Let each latest message have a score of 2^(-i) where i is the index of that latest message in the ranking.
+   *     Take a subset S of the latest messages such that the sum of scores is the greatest and
+   *     none of the blocks in S conflicts with each other. S will become the parents of the
+   *     about-to-be-created block.
+   *  3. Extract all valid deploys that aren't already in all ancestors of S (the parents).
+   *  4. Create a new block that contains the deploys from the previous step.
+   */
+  def create[F[_]: Sync: Log: Time: BlockStore: DeployStorage: Metrics: RuntimeManager: Span](
+      s: CasperSnapshot[F],
+      validatorIdentity: ValidatorIdentity
+      //dummyDeployOpt: Option[(PrivateKey, String)] = None
+  )(implicit runtimeManager: RuntimeManager[F]): F[BlockCreatorResult] =
+    Span[F].trace(ProcessDeploysAndCreateBlockMetricsSource) {
+      val selfId         = ByteString.copyFrom(validatorIdentity.publicKey.bytes)
+      val nextSeqNum     = s.maxSeqNums(selfId) + 1
+      val nextBlockNum   = s.maxBlockNum + 1
+      val parents        = s.parents
+      val justifications = s.justifications
+
+      def prepareUserDeploys(blockNumber: Long): F[Set[Signed[DeployData]]] =
+        for {
+          unfinalized         <- DeployStorage[F].getUnfinalized
+          earliestBlockNumber = blockNumber - s.onChainState.shardConf.deployLifespan
+          valid = unfinalized.filter(
+            d =>
+              notFutureDeploy(blockNumber, d.data) &&
+                notExpiredDeploy(earliestBlockNumber, d.data)
+          )
+          // this is required to prevent resending the same deploy several times by validator
+          validUnique = valid -- s.deploysInScope
+        } yield validUnique
+
+      def prepareSlashingDeploys(seqNum: Int): F[Seq[SlashDeploy]] =
+        for {
+          ilm <- s.dag.invalidLatestMessages
+          // if the node is already not active as per main parent, the node won't slash once more
+          ilmFromBonded = ilm.toList.filter {
+            case (validator, _) => s.onChainState.bondsMap.getOrElse(validator, 0L) > 0L
+          }
+          // TODO: Add `slashingDeploys` to DeployStorage
```

Comment:
I don't think this comment is relevant. Slashing deploys can always be generated when block is created so no need to store them.